File: Difficulty: Medium/Longest Substring with K Uniques/longest-substring-with-k-uniques.cpp
class Solution {
  public:
    int longestKSubstr(string &s, int k) {
        // code here
        int n = s.size();
        map<char,int> mp;
        int maxlen = -1;
        
        int l=0,r=0;
        
        while(r < n)
        {
            mp[s[r]]++;
            
            if(mp.size() > k)
            {
                mp[s[l]]--;
                
                if(mp[s[l]] == 0) mp.erase(s[l]);
                
                l++;
            }
            
            if(mp.size() == k) maxlen = max(maxlen,(r-l+1));
            
            r++;
        }
        
        
        
        return maxlen;
    }
};

File: Difficulty: Medium/Longest Substring with K Uniques/longest-substring-with-k-uniques.java
class Solution {
    public int longestKSubstr(String s, int k) {
        // code here
        int n = s.length();
        
        Map<Character,Integer> mp = new HashMap<>();
        
        int left = 0, right = 0;
        int maxLen = Integer.MIN_VALUE;
        
        while(right < n)
        {
            mp.put(s.charAt(right),mp.getOrDefault(s.charAt(right),0)+1);
            
            while(mp.size() > k)
            {
                mp.put(s.charAt(left),mp.getOrDefault(s.charAt(left),0)-1);
                
                if(mp.getOrDefault(s.charAt(left),0) == 0) 
                {
                    mp.remove(s.charAt(left));
                }
                
                left++;
            }
            
            if(mp.size() == k)
            {
                int len = right - left + 1;
                maxLen = Math.max(maxLen,len);
            }
            
            right++;
        }
        
        return maxLen == Integer.MIN_VALUE ? -1 : maxLen;
    }
}